### PR TITLE
Add Task as a built-in module/type

### PR DESCRIPTION
**Pull request by smores56**
This change moves the `Task` opaque IO type from platform-specific implementations to the standard library of Roc. This PR is ready for review, barring final testing by building a pre-release of `basic-cli` using the [twin basic-cli PR](https://github.com/roc-lang/basic-cli/pull/217). The high level changes are:

- Copied the `Task` module from the [basic-cli](https://github.com/roc-lang/basic-cli/blob/40e76f77af9ed6783b38a460dc7cb3b5d6cab885/platform/Task.roc) platform to `crates/compiler/builtins/roc/Task.roc`
- Consolidated `Task` and `Effect` into just the `Task` opaque type with definition `Task ok err := {} -> Result ok err`, which is the old type for `Effect a : {} -> a` that `Task ok err : Effect (Result ok err)` wrapped 
- Removed the `Effect` generation code, as everything should use `Task` directly
  - This includes the `generates Effect with [after, map, ...]` clause of the hosted module header, which is a breaking change
  - FFI definitions on the host side need to return `RocResult`s now, and FFI type declarations on the Roc side need to always have `Task` return types
  - Removed generation of the `Effect.after`, `Effect.map`, etc. functions as they were needed to glue `Effect` to `Task`, but are no longer needed now that those types have been combined
- Renamed hosted modules that contain FFI Roc declarations from `Effect.roc` to `PlatformTask.roc` to distinguish their behavior from the old approach and to give a name more descriptive of the new behavior
- Updated almost of the testing, but until `basic-cli` works, that is on hold

### CODE REVIEW

File: crates/compiler/can/src/task_module.rs
Comment (smores56) [inline crates/compiler/can/src/task_module.rs:18]: This needs to be removed before considering merging.

File: crates/compiler/can/src/task_module.rs
Comment (smores56) [inline crates/compiler/can/src/task_module.rs:173]: This needs to be removed before considering merging.

**Comment by lukewilliamboswell** [comment]
It looks like there is a genuine issue with the False interpreter. 

```
(nix:nix-shell-env) 192-168-1-103:false-interpreter luke$ roc build False.roc
🔨 Rebuilding platform...
0 errors and 0 warnings found in 940 ms
 while successfully building:

    False
(nix:nix-shell-env) 192-168-1-103:false-interpreter luke$ ./False examples/sqrt.false
Ran into problem:
examples/sqrt.false
```

**Comment by ageron** [comment]
`Task.sequence` runs tasks in reverse order, please see https://github.com/roc-lang/basic-cli/issues/235

**Comment by Anton-4** [comment]
TODO: rename `Task.seq` to `Task.sequence`. This was a recent change.

**Comment by Anton-4** [comment]
TODO from Richard: 
> I suspect this will be a lot faster by doing appends while building it up and then doing one reverse at the end

File: crates/compiler/builtins/roc/Task.roc
Comment (ayazhafiz) [inline crates/compiler/builtins/roc/Task.roc:39]: Why not remove this right now?

File: crates/compiler/builtins/roc/Task.roc
Comment (ayazhafiz) [inline crates/compiler/builtins/roc/Task.roc:77]:
```suggestion
## Consider the following task:
```

File: crates/compiler/builtins/roc/Task.roc
Comment (ayazhafiz) [inline crates/compiler/builtins/roc/Task.roc:81]:
```suggestion
## We can use [attempt] to handle the failure cases using the following:
```

File: crates/compiler/builtins/roc/Task.roc
Comment (ayazhafiz) [inline crates/compiler/builtins/roc/Task.roc:95]:
```suggestion
```

File: crates/compiler/builtins/roc/Task.roc
Comment (ayazhafiz) [inline crates/compiler/builtins/roc/Task.roc:113]: This example could be a bit better, for example to echo stdin to stdout. The current example doesn't use a material success value from the first task.

File: crates/compiler/can/src/def.rs
Comment (ayazhafiz) [inline crates/compiler/can/src/def.rs:2918]: wow lol this was a wild hack

File: crates/compiler/can/src/task_module.rs
Comment (ayazhafiz) [inline crates/compiler/can/src/task_module.rs:176]:
```suggestion
```

File: crates/compiler/builtins/roc/Task.roc
Comment (smores56) [reply crates/compiler/builtins/roc/Task.roc]: Done

File: crates/compiler/builtins/roc/Task.roc
Comment (smores56) [reply crates/compiler/builtins/roc/Task.roc]: Done

File: crates/compiler/builtins/roc/Task.roc
Comment (smores56) [reply crates/compiler/builtins/roc/Task.roc]: Good idea. I both made that the example, and mentioned how ! desugars to this method.

File: crates/compiler/can/src/def.rs
Comment (smores56) [reply crates/compiler/can/src/def.rs:2918]: Yeah... just another thing that results from codegen. The more we have in the std lib, the fewer hacks like this one that we'll need.

File: examples/cli/false-interpreter/platform/PlatformTask.roc
Comment (bhansconnect) [inline examples/cli/false-interpreter/platform/PlatformTask.roc:1]: Just curious, why is this named `PlatformTask`?


Could we maybe call this `Primitives`, `Tasks`, or `Effects`....maybe there is a better name? Cause `PlatformTask` really feels like it is defining a new `Task` type specific to the platform.

File: crates/cli/tests/benchmarks/platform/PlatformTasks.roc
Comment (Anton-4) [inline crates/cli/tests/benchmarks/platform/PlatformTasks.roc:9]: Should these be changed? [context](https://roc.zulipchat.com/#narrow/stream/316715-contributing/topic/Task.20as.20a.20built-in/near/465046819)

File: examples/cli/effects-platform/PlatformTasks.roc
Comment (Anton-4) [inline examples/cli/effects-platform/PlatformTasks.roc:7]: Same here

File: sort.roc
Comment (Anton-4) [inline sort.roc:1]: Should this file be here? It's at the root of the repo

File: crates/cli/tests/benchmarks/platform/PlatformTasks.roc
Comment (Anton-4) [reply crates/cli/tests/benchmarks/platform/PlatformTasks.roc:9]: Or should we just take care of that when we add the error generation for these return types?

**Review by Anton-4 — CHANGES_REQUESTED** [review]
The sort.roc file is the only thing that needs to be addressed/clarified, all good for the rest :tada:

File: sort.roc
Comment (Anton-4) [reply sort.roc:1]: Perhaps it's used as a script for something? In that case it would be good to add where it is used in a comment at the top of this file.

File: examples/cli/effects-platform/PlatformTasks.roc
Comment (smores56) [reply examples/cli/effects-platform/PlatformTasks.roc]: Used {} instead of *

File: crates/cli/tests/benchmarks/platform/PlatformTasks.roc
Comment (smores56) [reply crates/cli/tests/benchmarks/platform/PlatformTasks.roc]: Used {} instead of *

File: sort.roc
Comment (smores56) [reply sort.roc:1]: Removed, not sure why it was there.

**Review by Anton-4 — DISMISSED** [review]

**Comment by Anton-4** [comment]
Currently stuck on:
```
❯ ./target/release/roc examples/cli/false-interpreter/False.roc examples/cli/false-interpreter/examples/sqrt.false 
🔨 Rebuilding platform...
memory allocation of 93824993832009 bytes failed
```

**Comment by Anton-4** [comment]
Lol, the valgrind errors:
```
==217577== ERROR SUMMARY: 2963963 errors from 128 contexts (suppressed: 0 from 0)
```

**Comment by Anton-4** [comment]
I can continue investigating on Friday.
If anyone else wants to take a stab; I recommend stepping through the False executable with IDA free, side-by-side with the version on main.

**Comment by smores56** [comment]
We reduced all tests failures down to the same issue. After removing unsized tasks from the false-interpreter platform (the platform of the failing test), we still got a failure. Namely, the host returns `Ok(0)`, but Roc receives a `Err {}`, and we crash. The generated LLVM seems correct, so we are under the assumption that there is some lambda set issue here. The test has been therefore disabled for now, and we plan to revisit this once the lambda set fix is introduced.

If all tests pass, I vote that we call this good and merge, but I leave the final call to you @Anton-4.

**Comment by Anton-4** [comment]
All tests should pass now :crossed_fingers:

**Review by Anton-4 — APPROVED** [review]
:tada:

**Pull request merged**
